File: Toy-Language-Interpreter/DataTypes/MyDictionary.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Toy_Language_Interpreter.MyException;

namespace Toy_Language_Interpreter.DataTypes
{
    class MyDictionary<K, V> : MyIDictionary<K, V>
    {
        private Dictionary<K, V> dict;

        public MyDictionary()
        {
            this.dict = new Dictionary<K, V>();
        }

        public void Clone(MyIDictionary<K, V> toClone)
        {
            dict = new Dictionary<K, V>();
            foreach (K k in toClone.KeySet())
            {
                this.dict[k] = toClone.Get(k);
            }
        }

        public Dictionary<K, V> GetDict()
        {
            return this.dict;
        }

        public V Get(K key)
        {
            if (!dict.ContainsKey(key))
                throw new InvalidKeyMyDictionaryException("Invalid key: " + key.ToString());
            return dict[key];
        }

        public V Put(K key, V value)
        {
            dict[key] = value;
            return dict[key];
        }

        public bool Remove(K key)
        {
            if (!dict.ContainsKey(key))
                throw new InvalidKeyMyDictionaryException("Invalid key: " + key.ToString());
            return dict.Remove(key);
        }

        public bool ContainsKey(K key)
        {
            return dict.ContainsKey(key);
        }

        public bool ContainsValue(V value)
        {
            return dict.ContainsValue(value);
        }

        public List<K> KeySet()
        {
            return new List<K>(dict.Keys);
        }

        public List<V> Values()
        {
            return new List<V>(dict.Values);
        }

        public int Size()
        {
            return dict.Count;
        }

        public bool IsEmpty()
        {
            return dict.Count == 0;
        }

        public override string ToString()
        {
            String str = "";
            
            foreach(K k in dict.Keys)
            {
                str += k.ToString();
                str += " --> ";
                str += dict[k].ToString();
                str += "\n";
            }
            return str;
        }
    }
}

File: Toy-Language-Interpreter/src/DataTypes/MyDictionary.java
package DataTypes;

import java.io.Serializable;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import Exception.InvalidKeyMyDictionaryException;

/**
 * Created by vladc on 22.10.2016.
 */
public class MyDictionary<K, V> implements MyIDictionary<K, V>, Serializable {
    private Map<K, V> dict;

    public MyDictionary() {
        dict = new HashMap<>();
    }

    public void clone(MyIDictionary<K, V> toClone) {
        dict = new HashMap<>();
        for (K k : toClone.keySet()) {
            this.dict.put(k, toClone.getDict().get(k));
        }
    }

    public Map<K, V> getDict() {
        return this.dict;
    }

    public V get(Object key) throws InvalidKeyMyDictionaryException {
        if (!dict.containsKey(key))
            throw new InvalidKeyMyDictionaryException();
        return dict.get(key);
    }

    public V put(K key, V value) {
        if (!dict.containsKey(key))
            return dict.put(key, value);
        return dict.replace(key, value);
    }

    public V remove(Object key) throws InvalidKeyMyDictionaryException {
        if (!dict.containsKey(key))
            throw new InvalidKeyMyDictionaryException();
        return dict.remove(key);
    }

    public V replace(K key, V value) throws InvalidKeyMyDictionaryException {
        if (!dict.containsKey(key))
            throw new InvalidKeyMyDictionaryException();
        return dict.replace(key, value);
    }

    public boolean containsKey(Object key) {
        return dict.containsKey(key);
    }

    public boolean containsValue(Object value) {
        return dict.containsValue(value);
    }

    public Set<K> keySet() {
        return dict.keySet();
    }

    public Collection<V> values() {
        return dict.values();
    }

    public int size() {
        return dict.size();
    }

    public boolean isEmpty() {
        return dict.isEmpty();
    }

    @Override
    public String toString() {
        String str = "";
        for (K k : this.dict.keySet()) {
            str += k.toString();
            str += " --> ";
            str += this.dict.get(k).toString();
            str += "\n";
        }
        return str;
    }
}
